lte/gateway/python/magma/pipelined/ebpf/ebpf_dl_handler.c

```diff
@@ -0,0 +1,95 @@
+/**
+ * Copyright 2021 The Magma Authors.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <bcc/proto.h>
+#include <linux/in.h>
+#include <linux/ip.h>
+#include <linux/udp.h>
+#include <linux/if_ether.h>
+#include <uapi/linux/ipv6.h>
+
+// UE sessions map definitions.
+struct dl_map_key {
+  u32 ue_ip;
+};
+
+struct dl_map_info {
+  u32 remote_ipv4;
+  u32 tunnel_id;
+};
+
+// The map is pinned so that it can be accessed by pipelined or debugging tool
+// to examine datapath state.
+BPF_TABLE_PINNED(
+    "hash", struct dl_map_key, struct dl_map_info, dl_map, 1024 * 512,
+    "/sys/fs/bpf/dl_map");
+
+struct cfg_array_info {
+  u32 if_idx;
+};
+
+BPF_TABLE_PINNED(
+    "array", u32, struct cfg_array_info, cfg_array, 1, "/sys/fs/bpf/cfg_array");
+
+// Ingress handler for Uplink traffic.
+int gtpu_egress_handler(struct __sk_buff* skb) {
+  int ret;
+  struct iphdr* iph;
+  void* data;
+  void* data_end;
+
+  int ip_hdr = sizeof(struct ethhdr) + sizeof(struct iphdr);
+
+  data     = (void*) (long) skb->data;
+  data_end = (void*) (long) skb->data_end;
+  int len  = data_end - data;
+
+  // 1. a. check packet len
+  if ((data + ip_hdr) > data_end) {
+    bpf_trace_printk(
+        "ERR: truncated packet: len: %d, data sz %d\n", skb->len, len);
+    return TC_ACT_OK;
+  }
+
+  // 1. b. check UE map
+  iph                     = data + sizeof(struct ethhdr);
+  struct dl_map_key key   = {iph->daddr};
```

Comment:
key is overloaded here, lets call this dl_key or something better.

Reply:
done